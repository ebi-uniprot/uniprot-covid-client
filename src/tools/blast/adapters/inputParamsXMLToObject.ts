import { PublicServerParameters } from '../types/blastServerParameters';

// all the input parameters come from one endpoint, apart from the input
// sequence which comes from an other one, so separate them in the input of this
const inputParamsXMLToObject = (
  xmlString: string,
  sequence: string
): PublicServerParameters => {
  const doc = new window.DOMParser().parseFromString(xmlString, 'text/xml');
  const out: { [key: string]: string } = {};

  for (const node of doc.firstElementChild?.children || []) {
    let key;
    let value;
    for (const childNode of node.children) {
      switch (childNode.tagName) {
        case 'name':
          key = childNode.textContent;
          break;
        case 'value':
          value = childNode.textContent;
          break;
        default:
        // skip
      }
    }
    if (!(key && value)) {
      continue; // eslint-disable-line no-continue
    }
    out[key] = value.trim();
  }
  out.sequence = sequence;

  // 'as unknown as' to stop TypeScript from driving me mad ðŸ™„
  return (out as unknown) as PublicServerParameters;
};

export default inputParamsXMLToObject;
