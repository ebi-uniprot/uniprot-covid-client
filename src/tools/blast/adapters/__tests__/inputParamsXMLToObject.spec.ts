import fs from 'fs';
import path from 'path';
import { promisify } from 'util';

import inputParamsXMLToObject from '../inputParamsXMLToObject';

const readFile = promisify(fs.readFile);

let inputParams;
let inputParamsWithTaxons;
let sequence;

beforeAll(async () => {
  inputParams = await readFile(
    path.join(__dirname, '..', '__mocks__', 'input-params.xml'),
    'utf8'
  );
  inputParamsWithTaxons = await readFile(
    path.join(__dirname, '..', '__mocks__', 'input-params-with-taxons.xml'),
    'utf8'
  );
  sequence = await readFile(
    path.join(__dirname, '..', '__mocks__', 'input-sequence.fasta'),
    'utf8'
  );
});

describe('inputParamsXMLToObject tests', () => {
  it('should transform an XML string of input params to an object', () => {
    expect(inputParamsXMLToObject(inputParams, sequence)).toMatchSnapshot();
  });

  it('should transform an XML string of input params with taxons to an object', () => {
    expect(
      inputParamsXMLToObject(inputParamsWithTaxons, sequence)
    ).toMatchSnapshot();
  });
});
