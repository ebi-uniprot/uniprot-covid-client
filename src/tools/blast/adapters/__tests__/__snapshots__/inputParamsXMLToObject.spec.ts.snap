// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inputParamsXMLToObject tests should transform an XML string of input params to an object 1`] = `
Object {
  "align": "0",
  "alignments": "250",
  "compstats": "F",
  "database": "uniprotkb_refprotswissprot",
  "exp": "10",
  "filter": "F",
  "matrix": "BLOSUM62",
  "program": "blastp",
  "scores": "250",
  "sequence": ">sp|P06787|CALM_YEAST Calmodulin OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) OX=559292 PE=1 SV=1
MSSNLTEEQIAEFKEAFALFDKDNNGSISSSELATVMRSLGLSPSEAEVNDLMNEIDVDG
NHQIEFSEFLALMSRQLKSNDSEQELLEAFKVFDKNGDGLISAAELKHVLTSIGEKLTDA
EVDDMLREVSDGSGEINIQQFAALLSK",
  "stype": "protein",
}
`;

exports[`inputParamsXMLToObject tests should transform an XML string of input params with taxons to an object 1`] = `
Object {
  "align": "0",
  "alignments": "100",
  "compstats": "F",
  "database": "uniprotkb_refprotswissprot",
  "exp": "1e-1",
  "filter": "T",
  "gapalign": "true",
  "matrix": "BLOSUM45",
  "program": "blastp",
  "scores": "100",
  "sequence": ">sp|P06787|CALM_YEAST Calmodulin OS=Saccharomyces cerevisiae (strain ATCC 204508 / S288c) OX=559292 PE=1 SV=1
MSSNLTEEQIAEFKEAFALFDKDNNGSISSSELATVMRSLGLSPSEAEVNDLMNEIDVDG
NHQIEFSEFLALMSRQLKSNDSEQELLEAFKVFDKNGDGLISAAELKHVLTSIGEKLTDA
EVDDMLREVSDGSGEINIQQFAALLSK",
  "stype": "protein",
  "taxids": "9544,9606",
}
`;
